internal/extensionserver/post_translate_modify.go

```diff
@@ -490,3 +499,188 @@ func (s *Server) patchVirtualHostWithInferencePool(vh *routev3.VirtualHost, infe
+				})},
+			}
+
+			// Insert the AI Gateway extproc filter as the first extproc filter.
+			insertAIGatewayExtProcFilter(httpConManager, extProcFilter)
+			// Write the updated HCM back to the filter chain.
+			currChain.Filters[hcmIndex].ConfigType = &listenerv3.Filter_TypedConfig{TypedConfig: mustToAny(httpConManager)}
+		}
+	}
+}
+
+func (s *Server) isRouteGeneratedByAIGateway(route *routev3.Route) bool {
+	if s.isStandAloneMode {
+		// In stand-alone mode, we don't have the route metadata, so we assume that all routes are generated by AIGateway.
+		return true
+	}
+
+	// The route metadata should look like this:
+	//
+	//	filterMetadata:
+	//	 envoy-gateway:
+	//	   resources:
+	//	   - annotations:
+	//	       ai-gateway-generated: true
+	//      ....
+	//
+	// where "ai-gateway-generated" is the annotation that indicates that the route was generated by AIGateway.
+	// The below is to find it and enable the extproc filter for that route if it is found.
+	if route.Metadata == nil || route.Metadata.FilterMetadata == nil {
+		s.log.Info("no metadata found in the route, skipping", "route", route.Name)
+		return false
+	}
+
+	eg, ok := route.Metadata.FilterMetadata["envoy-gateway"]
+	if !ok {
+		s.log.Info("no envoy-gateway metadata found in the route, skipping", "route", route.Name)
+		return false
+	}
+	resources, ok := eg.Fields["resources"]
+	if !ok {
+		s.log.Info("no resources found in the envoy-gateway metadata, skipping", "route", route.Name)
+		return false
+	}
+	if resources.GetListValue() == nil || len(resources.GetListValue().Values) == 0 {
+		s.log.Info("no resources found in the envoy-gateway metadata, skipping", "route", route.Name)
+		return false
+	}
+
+	// Walk through the resources to find the AIGateway-generated HTTPRoute annotation.
+	for _, resource := range resources.GetListValue().Values {
+		if resource.GetStructValue() == nil {
+			s.log.Info("resource is not a struct, skipping", "route", route.Name, "resource", resource)
+			continue
+		}
+		annotations, ok := resource.GetStructValue().Fields["annotations"]
+		if !ok {
+			s.log.Info("no annotations found in the resource, skipping", "route", route.Name, "resource", resource)
+			continue
+		}
+		if annotations.GetStructValue() == nil {
+			s.log.Info("annotations is not a struct, skipping", "route", route.Name, "resource", resource)
+			continue
+		}
+		_, ok = annotations.GetStructValue().Fields[internalapi.AIGatewayGeneratedHTTPRouteAnnotation]
+		if ok {
+			return true
+		}
+	}
+	return false
+}
+
+func shouldAIGatewayExtProcBeInserted(filters []*httpconnectionmanagerv3.HttpFilter) bool {
+	for _, f := range filters {
+		if f.Name == aiGatewayExtProcName ||
+			// This is a backwards compatibility check for the old filter name generated by the EnvoyExtensionPolicy.
+			// https://github.com/envoyproxy/ai-gateway/blob/v0.2.1/internal/controller/gateway.go#L133
+			//
+			// TODO: remove after v0.3 release.
+			strings.Contains(f.Name, "ai-eg-eep-") {
```

Comment:
trying to understand this logic, when an old version of envoy ai gw upgraded, what will happen, will there be two aigw extproc? and if the old eep policy should be deleted directly